app/src/main/java/com/github/polybooks/database/FBInterestDatabase.kt

```diff
@@ -0,0 +1,212 @@
+
+        FirebaseProvider.getFirestore()
+            .collection(fieldCollection)
+            .get()
+            .addOnSuccessListener { documentSnapshots ->
+                val fieldsList = documentSnapshots.map { snapshot ->
+                    Field(name = snapshot.get("name") as String)
+                }
+                future.complete(fieldsList)
+            }
+            .addOnFailureListener { future.completeExceptionally(DatabaseException("Could not retrieve the list of all fields because of: $it")) }
+
+        return future
+    }
+
+
+    /**
+     * List all the Semesters in the database.
+     * */
+    override fun listAllSemesters(): CompletableFuture<List<Semester>> {
+        val future = CompletableFuture<List<Semester>>()
+
+        FirebaseProvider.getFirestore()
+            .collection(semesterCollection)
+            .get()
+            .addOnSuccessListener { documentSnapshots ->
+                val semestersList = documentSnapshots.map { snapshot ->
+                    Semester(
+                        section = snapshot.get("section") as String,
+                        semester = snapshot.get("semester") as String
+                    )
+                }
+                future.complete(semestersList)
+            }
+            .addOnFailureListener { future.completeExceptionally(DatabaseException("Could not retrieve the list of all semesters because of: $it")) }
+
+        return future
+    }
+
+    /**
+     * List all the Courses in the database.
+     * */
+    override fun listAllCourses(): CompletableFuture<List<Course>> {
+        val future = CompletableFuture<List<Course>>()
+
+        FirebaseProvider.getFirestore()
+            .collection(courseCollection)
+            .get()
+            .addOnSuccessListener { documentSnapshots ->
+                val coursesList = documentSnapshots.map { snapshot ->
+                    Course(name = snapshot.get("name") as String)
+                }
+                future.complete(coursesList)
+            }
+            .addOnFailureListener { future.completeExceptionally(DatabaseException("Could not retrieve the list of all courses because of: $it")) }
+
+        return future
+    }
+
+    /**
+     * Get the interests of the specified user
+     * TODO: Might need to add an authentication token to restrict authenticated users to only modify their interests.
+     * */
+    override fun getUserInterests(user: User): CompletableFuture<Triple<List<Field>, List<Semester>, List<Course>>> {
+        TODO("Not yet implemented")
+    }
+
+    /**
+     * Sets the interests of the specified user.
+     * TODO: Might need to add an authentication token to restrict authenticated users to only modify their interests.
+     * @return A Future to receive confirmation of success/failure asynchronously
+     * */
+    override fun setUserInterests(user: User, interests: List<Interest>): CompletableFuture<Unit> {
+        TODO("Not yet implemented")
+    }
```

Comment:
Should this also accept a list of semesters and list of courses (to mirror getUserInterests above)?

Reply:
We could make them more mirrored i.e. the input of one is the output of the other except for the future.

That being said, they're not fully incoherent now. Here it accepts a list of **interests** (any amongst field, semester and courses), while the other output 3 lists, one for each type of interest.

I think, above all, I'm keen on getting rid of this function and replace them by the addUserInterest / removeUserInterest.